File: bundle-plugin/src/main/scala/com/yahoo/container/plugin/mojo/GenerateOsgiManifestMojo.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.mojo

import java.io.File
import java.util.jar.{Attributes, JarEntry, JarFile}
import java.util.regex.Pattern

import com.yahoo.container.plugin.bundle.AnalyzeBundle
import com.yahoo.container.plugin.classanalysis.{Analyze, ExportPackageAnnotation, PackageTally}
import com.yahoo.container.plugin.mojo.GenerateOsgiManifestMojo._
import com.yahoo.container.plugin.osgi.ExportPackages
import com.yahoo.container.plugin.osgi.ExportPackages.Export
import com.yahoo.container.plugin.osgi.ImportPackages.Import
import com.yahoo.container.plugin.osgi.{ExportPackageParser, ExportPackages, ImportPackages}
import com.yahoo.container.plugin.util.Files.allDescendantFiles
import com.yahoo.container.plugin.util.IO.withFileOutputStream
import com.yahoo.container.plugin.util.Iteration.toStream
import com.yahoo.container.plugin.util.JarFiles.{withInputStream, withJarFile}
import com.yahoo.container.plugin.util.Strings
import org.apache.maven.artifact.Artifact
import org.apache.maven.plugin.{AbstractMojo, MojoExecutionException, MojoFailureException}
import org.apache.maven.plugins.annotations.{Mojo, Parameter, ResolutionScope}
import org.apache.maven.project.MavenProject

import scala.collection.immutable.Map


/**
 * @author  tonytv
 */
@Mojo(name = "generate-osgi-manifest", requiresDependencyResolution = ResolutionScope.TEST, threadSafe = true)
class GenerateOsgiManifestMojo extends AbstractMojo {

  @Parameter(defaultValue = "${project}")
  var project: MavenProject = null

  @Parameter
  var discApplicationClass: String = null

  @Parameter
  var discPreInstallBundle: String = null

  @Parameter(alias = "Bundle-Version", defaultValue = "${project.version}")
  var bundleVersion: String = null

  @Parameter(alias = "Bundle-SymbolicName", defaultValue = "${project.artifactId}")
  var bundleSymbolicName: String = null

  @Parameter(alias = "Bundle-Activator")
  var bundleActivator: String = null

  @Parameter(alias = "X-JDisc-Privileged-Activator")
  var jdiscPrivilegedActivator: String = null

  @Parameter(alias = "X-Config-Models")
  var configModels: String = null

  @Parameter(alias = "Import-Package")
  var importPackage: String = null

  @Parameter(alias = "WebInfUrl")
  var webInfUrl: String = null

  @Parameter(alias = "Main-Class")
  var mainClass: String = null

  @Parameter(alias = "X-Jersey-Binding")
  var jerseyBinding: String = null

  case class PackageInfo(name : String, exportAnnotation : Option[ExportPackageAnnotation])

  def execute() {
    try {
      val (jarArtifactsToInclude, jarArtifactsProvided, nonJarArtifacts) = Artifacts.getArtifacts(project)
      warnOnUnsupportedArtifacts(nonJarArtifacts)

      val publicPackagesFromProvidedJars = AnalyzeBundle.publicPackagesAggregated(jarArtifactsProvided.map(_.getFile))
      val includedJarPackageTally = definedPackages(jarArtifactsToInclude)

      val projectPackageTally = analyzeProjectClasses()

      val pluginPackageTally = projectPackageTally.combine(includedJarPackageTally)

      warnIfPackagesDefinedOverlapsGlobalPackages(projectPackageTally.definedPackages ++ includedJarPackageTally.definedPackages,
        publicPackagesFromProvidedJars.globals)

      if (getLog.isDebugEnabled) {
        getLog.debug("Referenced packages = " + pluginPackageTally.referencedPackages)
        getLog.debug("Defined packages = " + pluginPackageTally.definedPackages)
        getLog.debug("Exported packages of dependencies = " +
          publicPackagesFromProvidedJars.exports.map(e => (e.packageNames, e.version getOrElse "")).toSet )
      }

      val calculatedImports = ImportPackages.calculateImports(
        pluginPackageTally.referencedPackages,
        pluginPackageTally.definedPackages,
        ExportPackages.exportsByPackageName(publicPackagesFromProvidedJars.exports))

      val manualImports = emptyToNone(importPackage) map getManualImports getOrElse Map()

      createManifestFile(new File(project.getBuild.getOutputDirectory),
        manifestContent(
          project,
          jarArtifactsToInclude,
          manualImports,
          (calculatedImports -- manualImports.keys).values.toSet,
          pluginPackageTally))

    } catch {
      case e: MojoFailureException => throw e
      case e: MojoExecutionException => throw e
      case e: Exception => throw new MojoExecutionException("Failed generating osgi manifest.", e)
    }
  }

  //TODO: Tell which dependency overlaps
  private def warnIfPackagesDefinedOverlapsGlobalPackages(internalPackages: Set[String], globalPackages: List[String]) {
    val overlap = internalPackages intersect globalPackages.toSet
    if (overlap.nonEmpty)
      throw new MojoExecutionException(
        "The following packages are both global and included in the bundle:\n%s".format(overlap map ("   " + _) mkString ("\n")))
  }


  def osgiExportPackages(exportedPackages: Map[String, ExportPackageAnnotation]): Iterable[String] = {
    for ((name, annotation) <- exportedPackages)
      yield name + ";version=" + annotation.osgiVersion
  }

  def trimWhitespace(lines: Option[String]): String = {
    lines.getOrElse("").split(",").map(_.trim).mkString(",")
  }

  def manifestContent(project: MavenProject, jarArtifactsToInclude: Traversable[Artifact],
                      manualImports: Map[String, Option[String]], imports : Set[Import],
                       pluginPackageTally : PackageTally) = {
    Map[String, String](
      "Created-By" -> "vespa container maven plugin",
      "Bundle-ManifestVersion" -> "2",
      "Bundle-Name" -> project.getName,
      "Bundle-SymbolicName" -> bundleSymbolicName,
      "Bundle-Version" -> asBundleVersion(bundleVersion),
      "Bundle-Vendor" -> "Yahoo!",
      "Bundle-ClassPath" -> bundleClassPath(jarArtifactsToInclude),
      "Bundle-Activator" -> bundleActivator,
      "X-JDisc-Privileged-Activator" -> jdiscPrivilegedActivator,
      "Main-Class" -> mainClass,
      "X-JDisc-Application" -> discApplicationClass,
      "X-JDisc-Preinstall-Bundle" -> trimWhitespace(Option(discPreInstallBundle)),
      "X-Config-Models" -> configModels,
      "X-Jersey-Binding" -> jerseyBinding,
      "WebInfUrl" -> webInfUrl,
      "Import-Package" -> ((manualImports map asOsgiImport) ++ (imports map {_.asOsgiImport})).toList.sorted.mkString(","),
      "Export-Package" -> osgiExportPackages(pluginPackageTally.exportedPackages).toList.sorted.mkString(","))
      .filterNot  { case (key, value) => value == null || value.isEmpty }

  }

  def asOsgiImport(importSpec: (String, Option[String])) = importSpec match {
    case (packageName, Some(version)) => packageName + ";version=" + quote(version)
    case (packageName, None) => packageName
  }

  def quote(s: String) = '"' + s + '"'

  def createManifestFile(outputDirectory: File, manifestContent: Map[String, String]) {
    val manifest = toManifest(manifestContent)

    withFileOutputStream(new File(outputDirectory, JarFile.MANIFEST_NAME)) {
      outputStream =>
        manifest.write(outputStream)
    }
  }

  def toManifest(manifestContent: Map[String, String]) = {
    val manifest = new java.util.jar.Manifest
    val mainAttributes = manifest.getMainAttributes

    mainAttributes.put(Attributes.Name.MANIFEST_VERSION, "1.0")
    for ((key, value) <- manifestContent)
      mainAttributes.putValue(key, value)

    manifest
  }

  private def bundleClassPath(artifactsToInclude : Traversable[Artifact]) =
    ("." +: artifactsToInclude.map(dependencyPath).toList).mkString(",")

  private def dependencyPath(artifact : Artifact) =
    "dependencies/" + artifact.getFile.getName

  private def asBundleVersion(projectVersion: String) = {
    require(projectVersion != null, "Missing project version.")

    val parts = projectVersion.split(Pattern.quote("-"), 2)
    val numericPart = parts.head.split('.').map(Strings.emptyStringTo("0")).padTo(3, "0").toList

    val majorMinorMicro = numericPart take 3
    majorMinorMicro.mkString(".")
  }

  private def warnOnUnsupportedArtifacts(nonJarArtifacts: Traversable[Artifact]) {
    val unsupportedArtifacts = nonJarArtifacts.toSet.filter(_.getType != "pom")

    for (artifact <- unsupportedArtifacts) {
      getLog.warn(s"Unsupported artifact '${artifact.getId}': Type '${artifact.getType}' is not supported. Please file a feature request.")
    }
  }

  private def analyzeProjectClasses() : PackageTally = {
    val outputDirectory = new File(project.getBuild.getOutputDirectory)

    val analyzedClasses = allDescendantFiles(outputDirectory).filter(file => isClassToAnalyze(file.getName)).
      map(Analyze.analyzeClass)

    PackageTally.fromAnalyzedClassFiles(analyzedClasses)
  }

  def definedPackages(jarArtifacts: Iterable[Artifact]) : PackageTally = {
    PackageTally.combine(
      for (jarArtifact <- jarArtifacts) yield {
        withJarFile(jarArtifact.getFile) { jarFile =>
          definedPackages(jarFile)
        }
      })
  }

  def definedPackages(jarFile: JarFile) = {
    val analyzedClasses =
      for {
        entry <- toStream(jarFile.entries())
        if !entry.isDirectory
        if isClassToAnalyze(entry.getName)
        metaData = analyzeClass(jarFile, entry)
      } yield metaData

    PackageTally.fromAnalyzedClassFiles(analyzedClasses)
  }

  def analyzeClass(jarFile : JarFile, entry : JarEntry) = {
    try {
       withInputStream(jarFile, entry)(Analyze.analyzeClass)
    } catch {
      case e : Exception =>
        throw new MojoExecutionException(
          "While analyzing the class '%s' in jar file '%s'".format(entry.getName, jarFile.getName),
          e)
    }
  }
}

object GenerateOsgiManifestMojo {
  def getManualImports(importPackage: String): Map[String, Option[String]] = {
    try {
      (for {
        importDirective <- parseImportPackages(importPackage)
        packageName <- importDirective.packageNames
      } yield packageName -> getVersionThrowOthers(importDirective.parameters)).
        toMap

    } catch {
      case e: Exception => throw new RuntimeException("Error in Import-Package:" + importPackage, e)
    }
  }

  def getVersionThrowOthers(parameters: List[ExportPackages.Parameter]): Option[String] = {
    parameters match {
      case List() => None
      case List(ExportPackages.Parameter("version", v)) => Some(v)
      case default => throw new RuntimeException("A single, optional version parameter expected, but got " + default)
    }
  }

  def parseImportPackages(importPackages: String): List[Export] = {
    ExportPackageParser.parseAll(importPackages) match {
      case ExportPackageParser.NoSuccess(msg, _) => throw new RuntimeException(msg)
      case ExportPackageParser.Success(packages, _) => packages
    }
  }

  def isClassToAnalyze(name: String): Boolean =
    name.endsWith(".class") && ! name.endsWith("module-info.class")

  def emptyToNone(str: String) =
    Option(str) map {_.trim} filterNot  {_.isEmpty}
}

File: bundle-plugin/src/main/java/com/yahoo/container/plugin/mojo/GenerateOsgiManifestMojo.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.mojo;

import com.google.common.collect.Sets;
import com.yahoo.container.plugin.classanalysis.Analyze;
import com.yahoo.container.plugin.classanalysis.Analyze.JdkVersionCheck;
import com.yahoo.container.plugin.classanalysis.ClassFileMetaData;
import com.yahoo.container.plugin.classanalysis.PackageTally;
import com.yahoo.container.plugin.osgi.ExportPackages;
import com.yahoo.container.plugin.osgi.ExportPackages.Export;
import com.yahoo.container.plugin.osgi.ImportPackages.Import;
import com.yahoo.container.plugin.util.ArtifactId;
import com.yahoo.container.plugin.util.Artifacts;
import org.apache.maven.artifact.Artifact;
import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.apache.maven.plugins.annotations.ResolutionScope;

import java.io.File;
import java.nio.file.Paths;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.yahoo.container.plugin.bundle.AnalyzeBundle.exportedPackagesAggregated;
import static com.yahoo.container.plugin.bundle.AnalyzeBundle.nonPublicApiPackagesAggregated;
import static com.yahoo.container.plugin.classanalysis.Packages.disallowedImports;
import static com.yahoo.container.plugin.osgi.ExportPackages.exportsByPackageName;
import static com.yahoo.container.plugin.osgi.ImportPackages.calculateImports;
import static com.yahoo.container.plugin.util.Artifacts.VESPA_GROUP_ID;
import static com.yahoo.container.plugin.util.Artifacts.getVespaArtifact;
import static com.yahoo.container.plugin.util.Files.allDescendantFiles;
import static com.yahoo.container.plugin.util.JarFiles.providedArtifactsFromManifest;


/**
 * @author Tony Vaagenes
 * @author ollivir
 */
@Mojo(name = "generate-osgi-manifest", requiresDependencyResolution = ResolutionScope.TEST, threadSafe = true)
public class GenerateOsgiManifestMojo extends AbstractGenerateOsgiManifestMojo {

    private enum BundleType {
        CORE,      // up to container-dev
        INTERNAL,  // other vespa bundles (need not be set for groupId 'com.yahoo.vespa')
        USER
    }

    @Parameter
    private String discApplicationClass = null;

    @Parameter
    private String discPreInstallBundle = null;

    @Parameter(alias = "Bundle-Activator")
    private String bundleActivator = null;

    @Parameter(alias = "X-JDisc-Privileged-Activator")
    private String jdiscPrivilegedActivator = null;

    @Parameter(alias = "WebInfUrl")
    private String webInfUrl = null;

    @Parameter(alias = "Main-Class")
    private String mainClass = null;

    @Parameter(alias = "Bundle-Type")
    private BundleType bundleType = BundleType.USER;

    @Parameter(defaultValue = "false")
    private boolean suppressWarningMissingImportPackages;
    @Parameter(defaultValue = "false")
    private boolean suppressWarningPublicApi;
    @Parameter(defaultValue = "false")
    private boolean suppressWarningOverlappingPackages;
    @Parameter
    private List<String> allowEmbeddedArtifacts = List.of();

    @Parameter(defaultValue = "false")
    private boolean failOnWarnings;

    @Parameter(defaultValue = "false")
    private boolean buildLegacyVespaPlatformBundle;

    public void execute() throws MojoExecutionException {
        try {
            if (discPreInstallBundle != null && ! buildLegacyVespaPlatformBundle)
                throw new MojoExecutionException("The 'discPreInstallBundle' parameter can only be used by legacy Vespa platform bundles.");

            Artifacts.ArtifactSet artifactSet = Artifacts.getArtifacts(project);
            warnOnUnsupportedArtifacts(artifactSet.getNonJarArtifacts());

            List<Artifact> artifactsToInclude = artifactSet.getJarArtifactsToInclude();

            if (! isContainerDiscArtifact(project.getArtifact()))
                throwIfInternalContainerArtifactsAreIncluded(artifactsToInclude);

            List<Artifact> providedJarArtifacts = artifactSet.getJarArtifactsProvided();
            List<File> providedJarFiles = providedJarArtifacts.stream().map(Artifact::getFile).toList();
            List<Export> exportedPackagesFromProvidedJars = exportedPackagesAggregated(providedJarFiles);
            List<String> nonPublicApiPackagesFromProvidedJars = nonPublicApiPackagesAggregated(providedJarFiles);

            // Packages from Export-Package/PublicApi headers in provided scoped jars
            Set<String> exportedPackagesFromProvidedDeps = ExportPackages.packageNames(exportedPackagesFromProvidedJars);

            // Packaged defined in this project's code
            PackageTally projectPackages = getProjectClassesTally();

            // Packages defined in compile scoped jars
            PackageTally compileJarsPackages = definedPackages(artifactsToInclude);

            // The union of packages in the project and compile scoped jars
            PackageTally includedPackages = projectPackages.combine(compileJarsPackages);

            logDebugPackageSets(exportedPackagesFromProvidedJars, includedPackages);

            Optional<Artifact> jdisc_core = getVespaArtifact("jdisc_core", providedJarArtifacts);
            Optional<Artifact> wantedProvidedArtifact = getVespaArtifact(wantedProvidedDependency(), providedJarArtifacts);
            if (wantedProvidedArtifact.isPresent()) {
                // Having our wanted artifact as provided guarantees that log output does not contain its exported packages
                logMissingPackages(exportedPackagesFromProvidedDeps, projectPackages, compileJarsPackages, includedPackages);

                logProvidedArtifactsIncluded(artifactsToInclude, providedArtifactsFromManifest(wantedProvidedArtifact.get().getFile()));
            } else if (! suppressWarningMissingImportPackages && jdisc_core.isEmpty()) {
                // TODO: Remove jdisc_core clause above and instead add suppressWarning to necessary vespa modules.
                warnOrThrow(("This project does not have '%s' as provided dependency, so the generated 'Import-Package' " +
                        "OSGi header may be missing important packages.").formatted(wantedProvidedDependency()));
            }

            logOverlappingPackages(projectPackages, exportedPackagesFromProvidedDeps);

            Map<String, Export> exportedPackagesByName = exportsByPackageName(exportedPackagesFromProvidedJars);

            Map<String, Import> importsForProjectPackages = calculateImports(projectPackages.referencedPackages(),
                                                                             includedPackages.definedPackages(),
                                                                             exportedPackagesByName);
            List<String> nonPublicApiUsed = disallowedImports(importsForProjectPackages, nonPublicApiPackagesFromProvidedJars);
            logNonPublicApiUsage(nonPublicApiUsed);

            Map<String, Import> importsForIncludedPackages = calculateImports(includedPackages.referencedPackages(),
                                                                     includedPackages.definedPackages(),
                                                                     exportedPackagesByName);
            Map<String, String> manifestContent = generateManifestContent(artifactsToInclude, importsForIncludedPackages, includedPackages);
            addAdditionalManifestProperties(manifestContent);
            addManifestPropertiesForInternalAndCoreBundles(manifestContent, includedPackages, providedJarArtifacts);
            addManifestPropertiesForUserBundles(manifestContent, providedJarArtifacts, nonPublicApiUsed);

            createManifestFile(Paths.get(project.getBuild().getOutputDirectory()), manifestContent);
        } catch (Exception e) {
            throw new MojoExecutionException("Failed generating osgi manifest", e);
        }
    }

    private String wantedProvidedDependency() {
        return switch (effectiveBundleType()) {
            case CORE -> "jdisc_core";
            case INTERNAL -> "container-dev";
            case USER -> "container";
        };
    }

    private void addAdditionalManifestProperties(Map<String, String> manifestContent) {
        addIfNotEmpty(manifestContent, "Bundle-Activator", bundleActivator);
        addIfNotEmpty(manifestContent, "X-JDisc-Privileged-Activator", jdiscPrivilegedActivator);
        addIfNotEmpty(manifestContent, "Main-Class", mainClass);
        addIfNotEmpty(manifestContent, "X-JDisc-Application", discApplicationClass);
        addIfNotEmpty(manifestContent, "X-JDisc-Preinstall-Bundle", trimWhitespace(Optional.ofNullable(discPreInstallBundle)));
        addIfNotEmpty(manifestContent, "WebInfUrl", webInfUrl);
    }

    private void addManifestPropertiesForInternalAndCoreBundles(Map<String, String> manifestContent,
                                                                PackageTally includedPackages,
                                                                List<Artifact> providedJarArtifacts) {
        if (effectiveBundleType() == BundleType.USER) return;

        // TODO: this attribute is not necessary, remove?
        addIfNotEmpty(manifestContent, "X-JDisc-PublicApi-Package", publicApi(includedPackages));

        addIfNotEmpty(manifestContent, "X-JDisc-Non-PublicApi-Export-Package", nonPublicApi(includedPackages));
    }

    private void addManifestPropertiesForUserBundles(Map<String, String> manifestContent,
                                                     List<Artifact> providedArtifacts,
                                                     List<String> nonPublicApiUsed) {
        if (effectiveBundleType() != BundleType.USER) return;

        Optional<Artifact> jdisc_core = getVespaArtifact("jdisc_core", providedArtifacts);
        jdisc_core.ifPresent(
                artifact -> addIfNotEmpty(manifestContent, "X-JDisc-Vespa-Build-Version", artifact.getVersion()));
        addIfNotEmpty(manifestContent, "X-JDisc-Non-PublicApi-Import-Package", String.join(",", nonPublicApiUsed));
    }

    private void logNonPublicApiUsage(List<String> nonPublicApiUsed) {
        if (suppressWarningPublicApi || effectiveBundleType() != BundleType.USER || nonPublicApiUsed.isEmpty()) return;
        warnOrThrow("This project uses packages that are not part of Vespa's public api: %s".formatted(nonPublicApiUsed));
    }

    private static String publicApi(PackageTally tally) {
        return tally.publicApiPackages().stream().sorted().collect(Collectors.joining(","));
    }

    private static String nonPublicApi(PackageTally tally) {
        return tally.nonPublicApiExportedPackages().stream().sorted().collect(Collectors.joining(","));
    }

    private void logDebugPackageSets(List<Export> exportedPackagesFromProvidedJars, PackageTally includedPackages) {
        if (getLog().isDebugEnabled()) {
            getLog().debug("Referenced packages = " + includedPackages.referencedPackages());
            getLog().debug("Defined packages = " + includedPackages.definedPackages());
            getLog().debug("Exported packages of dependencies = " + exportedPackagesFromProvidedJars.stream()
                    .map(e -> "(" + e.getPackageNames().toString() + ", " + e.version().orElse("")).collect(Collectors.joining(", ")));
        }
    }

    private void logMissingPackages(Set<String> exportedPackagesFromProvidedJars,
                                    PackageTally projectPackages,
                                    PackageTally compileJarPackages,
                                    PackageTally includedPackages) {

        Set<String> definedAndExportedPackages = Sets.union(includedPackages.definedPackages(), exportedPackagesFromProvidedJars);

        Set<String> missingProjectPackages = projectPackages.referencedPackagesMissingFrom(definedAndExportedPackages);
        if (! missingProjectPackages.isEmpty()) {
            getLog().warn("Packages unavailable runtime are referenced from project classes " +
                                  "(annotations can usually be ignored): " + missingProjectPackages);
        }

        Set<String> missingCompilePackages = compileJarPackages.referencedPackagesMissingFrom(definedAndExportedPackages);
        if (! missingCompilePackages.isEmpty()) {
            getLog().info("Packages unavailable runtime are referenced from compile scoped jars " +
                                  "(annotations can usually be ignored): " + missingCompilePackages);
        }
    }

    private void logOverlappingPackages(PackageTally projectPackages,
                                        Set<String> exportedPackagesFromProvidedDeps) {
        if (suppressWarningOverlappingPackages) return;

        Set<String> overlappingProjectPackages = Sets.intersection(projectPackages.definedPackages(), exportedPackagesFromProvidedDeps);
        if (! overlappingProjectPackages.isEmpty()) {
            warnOrThrow("This project defines packages that are also defined in provided scoped dependencies " +
                         "(overlapping packages are strongly discouraged): " + overlappingProjectPackages);
        }
    }

    private void logProvidedArtifactsIncluded(List<Artifact> includedArtifacts,
                                              List<ArtifactId> providedArtifacts) throws MojoExecutionException {
        if (effectiveBundleType() == BundleType.CORE) return;

        Set<ArtifactId> included = includedArtifacts.stream().map(ArtifactId::fromArtifact).collect(Collectors.toSet());
        getLog().debug("Included  artifacts: " + included);
        getLog().debug("Provided  artifacts: " + providedArtifacts);

        Set<ArtifactId> includedProvided = Sets.intersection(included, new HashSet<>(providedArtifacts));
        getLog().debug("Included provided artifacts: " + includedProvided);
        HashSet<ArtifactId> allowed = getAllowedEmbeddedArtifacts(includedProvided);

        List<String> violations = includedProvided.stream()
                .filter(a -> ! allowed.contains(a))
                .map(ArtifactId::stringValue)
                .sorted().toList();

        if (! violations.isEmpty()) {
            warnOrThrow("Artifacts provided from Vespa runtime are included in compile scope: " + violations
                                + ". Direct dependencies should be removed."
                                + " For transitive dependencies, run 'mvn dependency:tree' and add necessary exclusions.");
        }
    }

    private HashSet<ArtifactId> getAllowedEmbeddedArtifacts(Set<ArtifactId> providedIncluded) throws MojoExecutionException {
        if (allowEmbeddedArtifacts.isEmpty()) return new HashSet<>();

        var allowed = new HashSet<ArtifactId>();
        try {
            allowEmbeddedArtifacts.stream().map(ArtifactId::fromStringValue).forEach(allowed::add);
        } catch (Exception e) {
            throw new MojoExecutionException("In config parameter 'allowEmbeddedArtifacts': " + e.getMessage(), e);
        }
        var allowedButUnused = Sets.difference(allowed, providedIncluded);
        if (! allowedButUnused.isEmpty()) {
            warnOrThrow("Configuration parameter 'allowEmbeddedArtifacts' contains artifact(s) not used in project: " + allowedButUnused);
        }
        getLog().info("Ignoring artifacts embedded in bundle: " + allowed);
        return allowed;
    }

    private static String trimWhitespace(Optional<String> lines) {
        return Stream.of(lines.orElse("").split(",")).map(String::trim).collect(Collectors.joining(","));
    }

    private void warnOnUnsupportedArtifacts(Collection<Artifact> nonJarArtifacts) {
        List<Artifact> unsupportedArtifacts = nonJarArtifacts.stream().filter(a -> ! a.getType().equals("pom"))
                .toList();

        unsupportedArtifacts.forEach(artifact -> warnOrThrow(String.format("Unsupported artifact '%s': Type '%s' is not supported. Please file a feature request.",
                                                                           artifact.getId(), artifact.getType())));
    }

    private void throwIfInternalContainerArtifactsAreIncluded(Collection<Artifact> includedArtifacts) throws MojoExecutionException {
        /* In most cases it's sufficient to test for 'component', as it's the lowest level container artifact,
         * Embedding container artifacts will cause class loading issues at runtime, because the classes will
         * not be equal to those seen by the framework (e.g. AbstractComponent). */
        if (includedArtifacts.stream().anyMatch(this::isJdiscComponentArtifact)) {
            throw new MojoExecutionException(
                    "This project includes the 'com.yahoo.vespa:component' artifact in compile scope." +
                            " It must have scope 'provided' to avoid resource leaks in your application at runtime." +
                            " Please use 'mvn dependency:tree' to find the root cause.");
        }
    }

    private BundleType effectiveBundleType() {
        if (bundleType != BundleType.USER) return bundleType;
        return isVespaInternalGroupId(project.getGroupId()) ? BundleType.INTERNAL : BundleType.USER;
    }

    private boolean isVespaInternalGroupId(String groupId) {
        return groupId.equals(VESPA_GROUP_ID)
                || groupId.equals(VESPA_GROUP_ID + ".hosted")
                || groupId.equals(VESPA_GROUP_ID + ".hosted.controller");
    }

    private boolean isJdiscComponentArtifact(Artifact a) {
        return a.getArtifactId().equals("component") && a.getGroupId().equals(VESPA_GROUP_ID);
    }

    private boolean isContainerDiscArtifact(Artifact a) {
        return a.getArtifactId().equals("container-disc") && a.getGroupId().equals(VESPA_GROUP_ID);
    }

    private PackageTally getProjectClassesTally() {
        File outputDirectory = new File(project.getBuild().getOutputDirectory());

        List<ClassFileMetaData> analyzedClasses = allDescendantFiles(outputDirectory)
                .filter(file -> file.getName().endsWith(".class"))
                .map(classFile -> Analyze.analyzeClass(classFile, JdkVersionCheck.ENABLED, artifactVersionOrNull(bundleVersion)))
                .toList();

        return PackageTally.fromAnalyzedClassFiles(analyzedClasses);
    }

    private void warnOrThrow(String... messages){
        String message = String.join("\n", messages);
        if (failOnWarnings) {
            throw new RuntimeException(message);
        }
        getLog().warn(message);
    }

}
